### PR TITLE
Remove BigArrays from SearchContext

### PR DESCRIPTION
**Pull request by nik9000**
We've been trying to shrink the big, mutable `SearchContext`. I'm doing
my part by removing `BigArrays` from it. Doing that required reworking
how we test `Aggregator`s to not need `SearchContext`. So I did that
too. Mostly. `top_hits` still needs a `SubSearchContext` which we can
still build, but it is now quite contained.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -768,8 +768,8 @@ private DefaultSearchContext createSearchContext(ReaderContext reader, ShardSear
         try {
             SearchShardTarget shardTarget = new SearchShardTarget(clusterService.localNode().getId(),
                 reader.indexShard().shardId(), request.getClusterAlias(), OriginalIndices.NONE);
-            searchContext = new DefaultSearchContext(reader, request, shardTarget, clusterService,
-                bigArrays, threadPool::relativeTimeInMillis, timeout, fetchPhase, lowLevelCancellation);
+            searchContext = new DefaultSearchContext(reader, request, shardTarget,
+                threadPool::relativeTimeInMillis, timeout, fetchPhase, lowLevelCancellation);
             // we clone the query shard context here just for rewriting otherwise we
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:773]: `clusterService` was unused so I removed it while I was clearing `bigArrays`.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregationBuilder.java [hunk 2]
```diff
@@ -401,11 +400,6 @@ public String getType() {
         return NAME;
     }
 
-    @Override
-    protected AggregationBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
-        return super.doRewrite(queryShardContext);
-    }
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregationBuilder.java:407]: This didn't do anything....

File: server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationContext.java [hunk 3]
```diff
@@ -230,6 +230,12 @@ public final AggregationUsageService getUsageService() {
      */
     public abstract Analyzer getIndexAnalyzer(Function<String, NamedAnalyzer> unindexedFieldAnalyzer);
 
+    /**
+     * Is this request cacheable? Requests that have
+     * non-deterministic queries or scripts aren't cachable.
+     */
+    public abstract boolean isCacheable();
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationContext.java:237]: These are required by tests. I could avoid adding it with some terrible casting stuff in tests but I think it is better to be up front about it.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationContext.java [hunk 4]
```diff
@@ -277,6 +284,7 @@ public ProductionAggregationContext(
             Supplier<Boolean> isCancelled
         ) {
             this.context = context;
+            this.bigArrays = context.bigArrays().withCircuitBreaking();  // We can break in searches.
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationContext.java:287]: Had to move with `withCircuitBreaking` from `SearchContext`.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 5]
```diff
@@ -297,25 +291,64 @@ public boolean shouldCache(Query query) {
...
+            query,
+            null,
+            consumer,
+            () -> buildSubSearchContext(mapperService, queryShardContext, bitsetFilterCache),
+            releasables::add,
+            bitsetFilterCache,
+            randomInt(),
+            () -> 0L,
+            () -> false
+        );
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:307]: I'd love to use a specialized testing subclass but this change is large enough as it is without me changing it.

File: test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java [hunk 6]
```diff
@@ -297,25 +291,64 @@ public boolean shouldCache(Query query) {
...
+            throw new RuntimeException(e);
+        }
+        when(ctx.fetchPhase()).thenReturn(new FetchPhase(Arrays.asList(new FetchSourcePhase(), new FetchDocValuesPhase())));
+        when(ctx.getQueryShardContext()).thenReturn(queryShardContext);
+        NestedDocuments nestedDocuments = new NestedDocuments(mapperService, bitsetFilterCache::getBitSetProducer);
+        when(ctx.getNestedDocuments()).thenReturn(nestedDocuments);
+        IndexShard indexShard = mock(IndexShard.class);
+        when(indexShard.shardId()).thenReturn(new ShardId("test", "test", 0));
+        when(ctx.indexShard()).thenReturn(indexShard);
+        return new SubSearchContext(ctx);
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:351]: Now all of this is contained to only be called for `top_hits`.

**Review by not-napoleon — APPROVED** [review]
LGTM, nice work

**Comment by nik9000** [comment]
> LGTM, nice work

Thanks for reviewing!

**Pull request merged**